### ISSUE TITLE
small/skinny matrix

### ISSUE THREAD

**Issue by AnonymousYWL** [flame/blis#521]
https://github.com/flame/blis/blob/master/docs/PerformanceSmall.md

Hello!
Is the evaluation of each small/skinny matrix start with cold data cache?

**Comment by devinamatthews** [comment]
I think they are hot cache. @fgvanzee ?

**Comment by fgvanzee** [comment]
Could somebody define these terms so I don't accidentally misspeak?

**Comment by fgvanzee** [comment]
@AnonymousYWL Okay, I realized that the definition doesn't matter; I can just explain what the test drivers do and then you will have the information to answer your question.

The driver starts by creating and initializing the matrices. It then enters a loop with a loop bound of 3. In the body of that loop, the `gemm` operation is performed. After the loop exits, all matrices are freed. The driver then terminates.

As so can see, we do not make any effort to dirty up the caches in between individual performance measurements. But that applies equally across all implementations.

**Comment by AnonymousYWL** [comment]
Is the same GEMM iterated multiple times?



 
---Original---
From: "Field G. Van ***@***.***&gt;
Date: 2021/7/8 01:14:02
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [flame/blis] small/skinny matrix (#521)




 
@AnonymousYWL Okay, I realized that the definition doesn't matter; I can just explain what the test drivers do and then you will have the information to answer your question.
 
The driver starts by creating and initializing the matrices. It then enters a loop with a loop bound of 3. In the body of that loop, the gemm operation is performed. After the loop exits, all matrices are freed. The driver then terminates.
 
As so can see, we do not make any effort to dirty up the caches in between individual performance measurements. But that applies equally across all implementations.

**Comment by fgvanzee** [comment]
> Is the same GEMM iterated multiple times?

Yes, three times.

**Comment by hominhquan** [comment]
@fgvanzee then the driver takes the average of the three runs I guess ? I think @AnonymousYWL is expecting something like:
```c
// cold cache
cold_start = bli_clock();
for (int i = 0; i < 1; i++) {
    bli_gemm_...(...);
}
cold_end = bli_clock();

// hot cache
hot_start = bli_clock();
for (int i = 0; i < 3 /*or positive number here*/; i++) {
    bli_gemm_...(...);
}
hot_end = bli_clock();
```

**Comment by fgvanzee** [comment]
> @fgvanzee then the driver takes the average of the three gemm I guess ?

No, as I said above, the driver always reports the best of three. (It is the max, not the average.) This applies for any implementation that the driver tests. 

> I think @AnonymousYWL is expecting something like:
> 
> ```c
> // cold cache
> cold_start = bli_clock();
> for (int i = 0; i < 1; i++) {
>     bli_gemm_...(...);
> }
> cold_end = bli_clock();
> 
> // hot cache
> hot_start = bli_clock();
> for (int i = 0; i < 3 /*or positive number here*/; i++) {
>     bli_gemm_...(...);
> }
> hot_end = bli_clock();
> ```

Perhaps, or perhaps they were merely trying to determine what is reported.

@AnonymousYWL Are you requesting that we *only* report performance under cold cache conditions?

**Comment by AnonymousYWL** [comment]
thanks for your reply.



I think hot cache will affect the performance of small matrices.
 
---Original---
From: "Minh Quan ***@***.***&gt;
Date: 2021/7/8 02:21:33
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [flame/blis] small/skinny matrix (#521)




 
@fgvanzee then the driver takes the average of the three runs I guess ? I think @AnonymousYWL is expecting something like:
 // cold cache cold_start = bli_clock(); for (int i = 0; i < 1; i++) {     bli_gemm_...(...); } cold_end = bli_clock(); // hot cache hot_start = bli_clock(); for (int i = 0; i < 3 /*or positive number here*/; i++) {     bli_gemm_...(...); } hot_end = bli_clock();

**Comment by AnonymousYWL** [comment]
Is it max GFLOPS or max clock?


 
---Original---
From: "Field G. Van ***@***.***&gt;
Date: 2021/7/8 02:26:14
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [flame/blis] small/skinny matrix (#521)




  
@fgvanzee then the driver takes the average of the three gemm I guess ?
  
No, as I said above, the driver always reports the best of three. (It is the max, not the average.) This applies for any implementation that the driver tests.
  
I think @AnonymousYWL is expecting something like:
 // cold cache cold_start = bli_clock(); for (int i = 0; i < 1; i++) {     bli_gemm_...(...); } cold_end = bli_clock(); // hot cache hot_start = bli_clock(); for (int i = 0; i < 3 /*or positive number here*/; i++) {     bli_gemm_...(...); } hot_end = bli_clock();
  
Perhaps, or perhaps they were merely trying to determine what is reported.
 
@AnonymousYWL Are you requesting that we only report performance under cold cache conditions?

**Comment by hominhquan** [comment]
> No, as I said above, the driver always reports the best of three. (It is the max, not the average.) This applies for any implementation that the driver tests.
Yes, I wrote this during your reply arrived, without refreshing the page.

@AnonymousYWL according to the testing way described by @fgvanzee, the best performance reported is probably the hot-cache performance. And we don't know if there is a big gap between cold and hot.

**Comment by AnonymousYWL** [comment]
Thank you for your reply.
Many thanks


 
---Original---
From: "Minh Quan ***@***.***&gt;
Date: 2021/7/8 02:32:15
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [flame/blis] small/skinny matrix (#521)




  
No, as I said above, the driver always reports the best of three. (It is the max, not the average.) This applies for any implementation that the driver tests.
 Yes, I wrote this during your reply arrived, without refreshing the page.
  
@AnonymousYWL according to the testing way described by @fgvanzee, the best performance reported is probably the hot-cache performance. And we don't know if there is a big gap between cold and hot.

**Comment by fgvanzee** [comment]
> Is it max GFLOPS or max clock?

Unfortunately, I'm not sure what you mean by "it."

We scale our y-axes according to the theoretical peak performance (measured in GFLOPS), which assumes the maximum attainable clock rate under the conditions of the experiment. (Sometimes clock rates are lower for multicore, for example, so we would not use the max clock rate for single-core in those cases.)

**Comment by devinamatthews** [comment]
TL;DR: yes it is hot cache.

@fgvanzee to get cold cache numbers we would have to clear the cache between runs. I don't think this is really necessary right now though. Maybe a note on the performance page though.
